### ISSUE TITLE
DateTimeInterface return type false positive

### ISSUE THREAD

**Issue by Slamdunk** [vimeo/psalm#419]
PHP DateTime/DateTimeImmutable natively return subtyped class, so the following code shouldn't generate errors.

---
Source code: https://3v4l.org/Q71qI
```php
final class MyDate extends DateTimeImmutable
{}

$foo = function (): MyDate {
    $today = new MyDate();
    $yesterday = $today->sub(new DateInterval('P1D'));

    return $yesterday;
};

/** @psalm-suppress ForbiddenCode */
var_dump(get_class($foo()));
// string(6) "MyDate"
```

---
Errors shown:
> **ERROR**: MoreSpecificReturnType - tmp.php:6:21 - The declared return type 'MyDate' for tmp.php:6:-:closure is more specific than the inferred return type 'DateTimeImmutable'
> ```php
> $foo = function (): MyDate {
>     $today = new MyDate();
>     $yesterday = $today->sub(new DateInterval('P1D'));
> 
>     return $yesterday;
> };
> ```
> 
> **ERROR**: LessSpecificReturnStatement - tmp.php:10:5 - The type 'DateTimeImmutable' is more general than the declared return type 'MyDate' for tmp.php:6:-:closure
> ```php
>     return $yesterday;
> ```

**Comment by muglug** [comment]
Thanks for finding! Should be fixed

**Comment by Slamdunk** [comment]
https://github.com/vimeo/psalm/commit/ff5b2ce79999b10213a8528ff5a2489473263de1 isn't enough unfortunately: the same issue goes for every other method of the same class and DateTime too: https://3v4l.org/1Mhfu

**Comment by muglug** [comment]
Sorry, I couldn't find that behaviour documented so I erred on the side of caution. I've made them all static in 24c69a9

**Comment by weirdan** [comment]
Not all of them behave like that. Static methods (`createFromFormat()`, `createFromMutable()`) **do not** return the subclass: https://3v4l.org/AaZIB

*Instance methods* of DateTime return the object itself, and instance methods*
